File: js/examples/swe/src/agents/inputs.ts
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import { ComposioToolSet } from "composio-core/lib/sdk/base.toolset";


type InputType = any;

function readUserInput(
  prompt: string,
  metavar: string,
  validator: (value: string) => InputType
): InputType {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise<InputType>((resolve, reject) => {
    rl.question(`${prompt} > `, (value) => {
      try {
        const validatedValue = validator(value);
        rl.close();
        resolve(validatedValue);
      } catch (e) {
        console.error(`Invalid value for \`${metavar}\` error parsing \`${value}\`; ${e}`);
        rl.close();
        reject(e);
      }
    });
  });
}

function githubRepositoryNameValidator(name: string): [string, string] {
  if (name.includes(' ')) {
    throw new Error();
  }
  const [owner, repoName] = name.split('/');
  return [owner, repoName];
}

function createGithubIssueValidator(owner: string, name: string, toolset: ComposioToolSet) {
  return async function githubIssueValidator(value: string): Promise<string> {
    const resolvedPath = path.resolve(value);
    if (fs.existsSync(resolvedPath)) {
      return fs.readFileSync(resolvedPath, 'utf-8');
    }

    if (/^\d+$/.test(value)) {
      const responseData = await toolset.executeAction('github_issues_get', {
        owner,
        repo: name,
        issue_number: parseInt(value, 10),
      });
      return responseData.body as string;
    }

    return value;
  };
}

export async function fromGithub(toolset: ComposioToolSet): Promise<{ repo: string; issue: string }> {
  const owner = await readUserInput(
    'Enter github repository owner',
    'github repository owner',
    (value: string) => value
  );
  const name = await readUserInput(
    'Enter github repository name',
    'github repository name',
    githubRepositoryNameValidator
  );
  const repo = `${owner}/${name}`;
  const issue = await readUserInput(
    'Enter github issue ID or description or path to the file containing description',
    'github issue',
    createGithubIssueValidator(owner, name, toolset)
  );
  return { repo, issue };
}

File: python/swe/agent/inputs.py
import re
import typing as t
from pathlib import Path
import os

from composio import App, Action, ComposioToolSet


InputType = t.TypeVar("InputType")
workspace_dir = "/Users/shrey/trial_repos/"
composio_toolset = ComposioToolSet()

def read_user_input(
    prompt: str,
    metavar: str,
    validator: t.Callable[[str], InputType],
) -> InputType:
    """Read user input."""
    while True:
        value = input(f"{prompt} > ")
        try:
            return validator(value)
        except Exception as e:
            print(f"Invalid value for `{metavar}` error parsing `{value}`; {e}")


def _github_repository_name_validator(name: str) -> t.Tuple[str, str]:
    """Validate github repository name."""
    if " " in name:
        raise ValueError()
    owner, name = name.split("/")
    return owner, name


def _create_github_issue_validator(owner: str, name: str) -> t.Callable[[str], str]:
    """Create a github issue validator."""

    def _github_issue_validator(value: str) -> str:
        """Validate github issue."""
        if Path(value).resolve().exists():
            return Path(value).read_text(encoding="utf-8")

        if re.match(r"^\d+$", value):
            response_data = composio_toolset.execute_action(
                action=Action.GITHUB_GET_AN_ISSUE,
                params={
                    "owner": owner,
                    "repo": name,
                    "issue_number": int(value),
                },
            ).get("response_data")
            return response_data["body"]

        return value

    return _github_issue_validator


def set_workspace() -> t.Tuple[str, str, str]:
    """Take input from github."""
    # owner, name = read_user_input(
    #     prompt="Enter github repository name",
    #     metavar="github repository name",
    #     validator=_github_repository_name_validator,
    # )
    # owner, name = "pallets", "flask"
    owner, name = "shreysingla11", "Aip-projecct"

    # base_commit = read_user_input(
    #     prompt="Enter base commit id, leave empty for default branch",
    #     metavar="base commit id",
    #     validator=lambda x: x,
    # )

    # base_commit = "7ee9ceb71e868944a46e1ff00b506772a53a4f1d"
    base_commit = ""

    composio_toolset.execute_action(
        action=Action.FILETOOL_CHANGE_WORKING_DIRECTORY,
        params={"path": workspace_dir},
    )

    composio_toolset.execute_action(
        action=Action.FILETOOL_GIT_CLONE,
        params={
            "repo_name": f"{owner}/{name}",
            "commit_id": base_commit,
        },
    )
    
    return (
        f"{owner}/{name}",
        str(Path(workspace_dir) / name),
        # read_user_input(
        #     prompt="Enter github issue ID or description or path to the file containing description",
        #     metavar="github issue",
        #     validator=_create_github_issue_validator(
        #         owner=owner,
        #         name=name,
        #     ),
        # ),
        # "Require a non-empty name for Blueprints Things do not work correctly if a Blueprint is given an empty name (e.g. #4944). It would be helpful if a `ValueError` was raised when trying to do that."
        "In the file `code_base/classify_base_on_features.py`, improve the implementation of the 'classify_n_features' function."
    )
